docs/specs/captcha.md

```diff
@@ -1,41 +1,354 @@
+    + [reCAPTCHA Enterprise](#recaptcha-enterprise)
+    + [Cloudflare Turnstile](#cloudflare-turnstile)
+    + [Arkose Labs Bot Manager](#arkose-labs-bot-manager)
+    + [Tencent Captcha](#tencent-captcha)
 
-## Configurations
+## Old configuration
 
-### authgear.yaml
+> The old configuration **IS NOT** used by Authentication Flow.
+> To configure Captcha providers for an Authentication Flow, the new configuration must be used.
+
+See [here](./captcha_legacy.md#configuration) for the documentation of the old configuration.
 
-Before using any workflow with captcha enabled, the captcha provider must be configured in `authgear.yaml`:
+## New Configuration
+
+The section documents the new configuration.
+
+### authgear.yaml
 
 ```yaml
 captcha:
-  provider: cloudflare
+  enabled: true
+  providers:
+  - type: cloudflare
+    alias: cloudflare
+    site_key: "SITE_KEY"
 ```
 
-- `provider`
-  - The captcha provider. Currently, only `cloudflare` is supported.
+- `captcha.enabled`: Boolean. If it is true, the new configuration is used.
+- `captcha.providers`: An array of Captcha provider configuration. The actual shape depends on the `type` property.
+- `captcha.providers.type`: Required. The type of the Captcha provider. Valid values are `cloudflare` and `recaptchav2`.
+- `captcha.providers.alias`: Required. The unique identifier to the Captcha provider. It is used in other parts of the configuration to refer to this particular Captcha provider. For example, the project can configured a number of Captcha providers, but only uses one of them in a particular Authentication Flow.
 
-### authgear.secrets.yaml
+Other fields are specific to `type`.
+
+#### `type: cloudflare`
 
-When the captcha provider is configured, the corresponding secret must exist in `authgear.secrets.yaml`.
+- `site_key`: The site key of Cloudflare Turnstile.
+
+#### `type: recaptchav2`
+
+- `site_key`: The site key of reCAPTCHA v2.
+
+### authgear.secrets.yaml
 
 ```yaml
 - data:
-    secret: YOUR_TURNSTILE_SECRET_KEY
-  key: captcha.cloudflare
+    items:
+    - type: cloudflare
+      alias: cloudflare
+      secret: TURNSTILE_SECRET_KEY
+  key: captcha.providers
+```
+
+- `items.type`: Required. It is the same as `captcha.providers.type`.
+- `items.alias`: Required. It is the same as `captcha.providers.alias`.
+
+Other fields are specific to `type`.
+
+#### `type: cloudflare`
+
+- `secret_key`: Required. The secret key of Cloudflare Turnstile.
+
+#### `type: recaptchav2`
+
+- `secret_key`: Required. The secret key of reCAPTCHA v2.
+
+## Authentication Flow
+
+This section specifies how Captcha works in a Authentication Flow.
+
+### Captcha in Authentication Flow configuration
+
+Captcha is supported in the following flow types:
+
+- `signup`
+- `promote`
+- `login`
+- `signup_login`
+- `reauth`
+- `account_recovery`
+
+Captcha is supported only in the following step types:
+
+- `identify` in `signup`, `promote`, `login`, `signup_login`, and `account_recovery`.
```

Comment:
Is `reauth` intentionally left out here? 🙏

Reply:
The `identify` step in `reauth` is done by an ID token, which involves no user interaction. So the captcha is supported in `authenticate` instead.